### ISSUE TITLE
手动替换lora问题

### ISSUE THREAD

**Issue by POSE200** [jiangsongtao/TinyMed#32]
作者，你好，这是我训练是的手动替换lora，使用的是你提供的lora_moe.py,在train.py的1320行左右，跳过了初始化MoE类的初始化，使用lora初始化和权重加载，因为lora_moe.py里面似乎有相应的逻辑，之后我在评估是发现出现错误，因为调用了evalMoEStable那个类，他是直接初始化的，但是lora_moe.py版本缺少[num_expet]键，所以我换成了用MoESTablelm这个训练是的类，初始化，只不过training设置为False，在buider.py的130行左右，

    training_args.moe_enable = model_args.moe_enable
    training_args.only_lora_ffn = model_args.only_lora_ffn
    model_args.lora_enable = training_args.lora_enable
    def apply_lora_moe_to_model(model, model_args, training_args):
        """将LoRA_MOE模块应用到模型中"""
        from moe.lora_moe import LoRA_MOE_LM2
        model.router_aux_loss_coef = getattr(model_args, 'router_aux_loss_coef', 0.0)

        lora_applied_count = 0

        # 遍历模型的所有MLP层
        for name, module in model.named_modules():
            if (name.endswith('.mlp') and
                hasattr(module, 'gate_proj') and
                hasattr(module, 'up_proj') and
                hasattr(module, 'down_proj')):

                print(f"找到MLP层: {name}")
                parent_module = model
                for part in name.split('.')[:-1]:
                    parent_module = getattr(parent_module, part)
                lora_moe = LoRA_MOE_LM2(
                    lora_rank=training_args.lora_r,
                    lora_alpha=training_args.lora_alpha,
                    num_experts=4,
                    training=True,
                    original_module=module
                )

                # 替换原始MLP为LoRA_MOE
                layer_name = name.split('.')[-1]
                setattr(parent_module, layer_name, lora_moe)

                lora_applied_count += 1
                print(f"成功替换 {name} 为 LoRA_MOE")

        print(f"总共替换了 {lora_applied_count} 个MLP层为LoRA_MOE")
        return model
    if model_args.moe_enable:
        if training_args.lora_enable:
            # 使用自定义LoRA_MOE模块
            rank0_print("Applying custom LoRA_MOE modules...")
            model = apply_lora_moe_to_model(model, model_args, training_args)
            rank0_print("LoRA_MOE modules applied successfully!")
        else:
            model.initialize_moe_modules(model_args=model_args)
    

 buider.py      170行左右
他这里有lora识别模型名称的逻辑，但是似乎没有使用，里面没有lora逻辑，所以我就加了初始化，加载评估是的lora模型

      elif 'lora' in model_name.lower() and 'moe' in model_name.lower():
                  lora_cfg_pretrained = AutoConfig.from_pretrained(model_path)
                  print('Adapting to MoE...')
                  if 'qwen' in model_name.lower() and '1.5' not in model_name.lower():
                      tokenizer = QWenTokenizer.from_pretrained(model_path, use_fast=False, padding_side=padding_side)
                      model = EvalMoELLaVAQWenForCausalLM.from_pretrained(model_path, low_cpu_mem_usage=True, config=lora_cfg_pretrained, **kwargs)
                      model.generation_config = GenerationConfig.from_pretrained(model_path, pad_token_id=tokenizer.pad_token_id)
                      # model.generation_config.repetition_penalty = None
                      model.generation_config.do_sample = False  # use greedy decoding
                      model.generation_config.repetition_penalty = 1.0  # disable repetition penalty
                  elif 'openchat' in model_name.lower() or 'mistral' in model_name.lower():
                      tokenizer = AutoTokenizer.from_pretrained(model_path, use_fast=False, padding_side=padding_side)
                      model = EvalMoELLaVAMistralForCausalLM.from_pretrained(model_path, low_cpu_mem_usage=True, config=lora_cfg_pretrained, **kwargs)
                  elif 'qwen' in model_name.lower() and '1.5' in model_name.lower():
                      tokenizer = AutoTokenizer.from_pretrained(model_path, use_fast=False, padding_side=padding_side)
                      model = EvalMoELLaVAQwen1_5ForCausalLM.from_pretrained(model_path, low_cpu_mem_usage=True, **kwargs)
                      # import ipdb
                      # ipdb.set_trace()
                      model.config.eos_token_id = tokenizer.eos_token_id
                  elif 'phi' in model_name.lower():
                      tokenizer = AutoTokenizer.from_pretrained(model_path, use_fast=False, padding_side=padding_side)
                      model = EvalMoELLaVAPhiForCausalLM.from_pretrained(model_path, low_cpu_mem_usage=True, config=lora_cfg_pretrained, **kwargs)
                      model.config.eos_token_id = tokenizer.eos_token_id
                  elif 'minicpm' in model_name.lower():
                      tokenizer = AutoTokenizer.from_pretrained(model_path, use_fast=False, padding_side=padding_side)
                      model = EvalMoELLaVAMiniCPMForCausalLM.from_pretrained(model_path, low_cpu_mem_usage=True, config=lora_cfg_pretrained, **kwargs)
                      model.config.eos_token_id = tokenizer.eos_token_id
                  elif 'stablelm' in model_name.lower():
                      from moellava.model.language_model.stablelm.tokenization_arcade100k import Arcade100kTokenizer
                      #tokenizer = Arcade100kTokenizer.from_pretrained(model_path, use_fast=False, padding_side=padding_side)
                      tokenizer = AutoTokenizer.from_pretrained(model_path, use_fast=False, padding_side=padding_side)
                      model = MoELLaVAStablelmForCausalLM.from_pretrained(model_path, low_cpu_mem_usage=True,trust_remote_code=True,config=lora_cfg_pretrained, **kwargs)
                      print("Applying custom LoRA-MoE to loaded EvalMoELLaVAStablelmForCausalLM...")
                      from moe.lora_moe import LoRA_MOE_LM2
      
                      lora_applied_count = 0
                      for name, module in model.named_modules():
                          if (name.endswith('.mlp') and hasattr(module, 'gate_proj') and
                              hasattr(module, 'up_proj') and hasattr(module, 'down_proj')):
      
                              parent_module = model
                              for part in name.split('.')[:-1]:
                                  parent_module = getattr(parent_module, part)
      
                              lora_moe = LoRA_MOE_LM2(
                                  lora_rank=128, lora_alpha=256, num_experts=4,
                                  training=False, original_module=module)
      
                              layer_name = name.split('.')[-1]
                              setattr(parent_module, layer_name, lora_moe)
                              lora_applied_count += 1
                              print(f"Replaced {name} with LoRA-MoE")
      
                      print(f"Total LoRA-MoE replacements: {lora_applied_count}")
      
                      model.config.eos_token_id = tokenizer.eos_token_id

    期间有一些小错误，但是都无伤大雅，就是因为router_loss_coef=0所以我注释了这个逻辑，因为有他会报错

    # if len(outputs[-1]) > 0:
        #     moe_loss_list = outputs[-1]
            # import ipdb
            # ipdb.set_trace()
            # for moe_loss in moe_loss_list:
            #     if moe_loss is not None:
            #         moe_losses.append(moe_loss)
            # router_coef = getattr(self, 'router_aux_loss_coef', 0.01)
            # moe_loss = router_coef * sum(moe_losses)
            # if labels is not None:
            #     print(loss, sum(moe_losses), loss + moe_loss)
            #     loss += moe_loss
        # import ipdb
        # ipdb.set_trace()
在llava_stablelm_moe.py的390行左右 
训练完加载逻辑后，结果评估是全是乱码，请问我的做法对吗？

**Comment by jiangsongtao** [comment]
你好,不好意思上个礼拜在赶ddl, 我觉得可能是merge的时候有一些权重key没对上,导致权重出现了NAN, 我建议你可以check一下, 可以先merge好再去推理, 具体可以参考一下这个项目,https://github.com/forwchen/LLaVA-MoLE/blob/5876301bc4357b91ec45c7d3d475cde958eb3593/moe.py,是使用lora来训练多模态moe的,应该大体逻辑会是一样,我的moelora的实现也是参考他的

**Comment by POSE200** [comment]
作者，你好，那如果完整训练出来权重文件是多少G，如果是lora模式，我的是4.3G，如果不对，估计就是训练完保存的权重不对。而我看不用lora的权重是7G往上。

**Comment by POSE200** [comment]
你好，作者，还有想问一下，他如果使用lora，这里的切片数是2个还是三个，我如果运行PEFT库这个有bug的lora替换是[2,2,2]，三个切片显示，如果是手动替换的是[2,2]，这样方便我快速定位问题，谢谢

<img width="1122" height="58" alt="Image" src="https://github.com/user-attachments/assets/0d1807bc-3567-4658-b241-75ef5cc86397" />

**Comment by jiangsongtao** [comment]
你好,

<img width="857" height="176" alt="Image" src="https://github.com/user-attachments/assets/858c211f-d0d5-4868-b84e-9a02a0937c0d" />我跑的显示是这个哈

**Comment by POSE200** [comment]
作者，那能提供一下你在原码里面替换lora的关键文件吗，比如说train.py这些文件配置的手写lora，因为我发现我的参数量异常。

<img width="1727" height="136" alt="Image" src="https://github.com/user-attachments/assets/3d120b6b-726b-47f3-8994-e1bc048eb991" />

**Comment by jiangsongtao** [comment]
你好，替换文件就是我之前发的那个lora_moe哈，把之前llava_stablelm_moe.py里面替换deepspped的moe的部分换成lora_moe里面的moe就行，train.py应该不需要更改，也不需要设置lora_enable=True，如果爆显存你可以把lora的秩调小一点

**Comment by POSE200** [comment]
你好作者我发现全量微调比lora还要差，怎么回事呢，第一张全量，第二张lora，

<img width="311" height="210" alt="Image" src="https://github.com/user-attachments/assets/c9ec41cd-623e-4304-9a06-c6e091993c90" />

<img width="301" height="223" alt="Image" src="https://github.com/user-attachments/assets/8723a99a-7c65-4c9c-b3bd-3ddc17a1591d" />

**Comment by jiangsongtao** [comment]
你好，请问你训练使用的数据集是什么？以及训练的epoch，之前实验的经验是全量微调如果训练数据只是某几个vqa的话，并且epoch少的话效果可能不如lora，加大epoch，全量微调在vqa上的性能表现会好于lora, 但是模型可能存在过拟合的情况，因此我建议你可以混合一部分指令微调数据进来

**Comment by jiangsongtao** [comment]
你好，还有就是请问你最后成功运行了嘛，替换成loramoe.py里面的moe后

**Comment by POSE200** [comment]
作者你好我用的数据集是，就是你提供的第三阶段标准数据集

<img width="962" height="88" alt="Image" src="https://github.com/user-attachments/assets/42a6ba69-ff29-4744-a7cd-489fe6f4ba26" />
还有全量和lora我都使用了9个epoch和原码一样，而对于替换手写lora，我已经成功替换并运行了，替换逻辑是这样的，请问作者一样吗？


      for num_experts, layer_num in zip(self.config.moe['num_experts'], moe_layers_idx):
                pretrained_state_dict = self.model.layers[layer_num].mlp.state_dict()
                original_mlp = self.model.layers[layer_num].mlp
                from moe.lora_moe import LoRA_MOE_LM2
                self.model.layers[layer_num].mlp = LoRA_MOE_LM2(
                    lora_rank=getattr(model_args, 'lora_rank', 128),
                    lora_alpha=getattr(model_args, 'lora_alpha', 256),
                    num_experts=num_experts,
                    training=True,
                    original_module=original_mlp
              )


并且全量我用的还是双卡，lora是单卡

**Comment by jiangsongtao** [comment]
你好，看起来你的设置没有问题，我之前lora rank在256的时候，性能确实和全量微调差距很小了，不同机器不同环境下，训练出来的lora也可能会比全量要好，你也可以save_steps设置一下，看看在3,6,9 epoch下的性能比较. 如果stage3只做vqa的话其实我认为使用lora会更好，如果你在stage3会加入更多指令微调数据集例如huatuogpt-vision等等，我建议使用全量微调会更好，现在VLMs一般不会再单独针对vqa进行一个stage的训练了，在去年的时候我们是follow的LLava-Med的这个setting, 后面发现把vqa mixed进指令微调数据集训练出来的模型会更加实用，不然如果只在vqa上微调，模型可能只会回答VQA. 因此我建议你可以根据自己的任务来定哈.

**Comment by POSE200** [comment]
好的，感谢作者的解释

**Comment by jiangsongtao** [comment]
不客气哈~, 很感谢你对我们项目感兴趣, 也很开心能够帮助到你~
